### DISCUSSION TITLE
Improve Pandoc's syntax highlighting output (Skylighting tokens and span styles are a bit too simplistic)

### DISCUSSION THREAD

**Discussion by badlydrawnrob** [jgm/pandoc#10115] [General]
> **TL;DR:** Skylighting makes some themes impossible to recreate.
> The tokeniser (I think that's the right term) is more simplistic than other highlighting tools, and possibly wrong in some areas.

I've searched for similar issues but there are _a lot_ of syntax highlighting (closed) issues! **I've made [some observations here](https://github.com/badlydrawnrob/anki/issues/152)** about how Pandoc is outputting it's syntax highlighting for Markdown fenced code blocks, as I've been running into some roadblocks when trying to create my own colour themes.

I understand you're using Haskell's Skylighting. I'm not a Haskell guy (I've seen [a couple](https://vaibhavsagar.com/blog/2023/01/29/ghc-syntax-hakyll/) of [workarounds](https://tony-zorman.com/posts/pygmentising-hakyll.html) but I'm not a good enough programmer to understand them!

- **Are there any known issues of this sort, or future plans for improving the highlighting** to get closer to Visual Studio Code's, [highlight.js](https://highlightjs.org/), or [Pygments](https://pygments.org/) (for example) highlighting capabilities?
- **Are there any simple ways to extend Skylighting that a novice programmer could achieve?** (or choose a different highlighting tool for Pandoc to use)
    - Or at least advise if [one of the fixes](https://vaibhavsagar.com/blog/2023/01/29/ghc-syntax-hakyll/) I've mentioned above are safe to implement.

Pandoc, in general, works just fine and is easy to use and compile for my purposes, but some languages (like Elm Lang) render quite poorly, for example, `.fu` is applied to both functions and it's arguments, so the [Breeze Dark Theme](https://marketplace.visualstudio.com/items?itemName=AndrewFridley.Breeze-Dark-Theme) is impossible to recreate:

<img width="605" alt="model-div-span-01-correct-colours" src="https://github.com/user-attachments/assets/3dcf4b66-78ec-46e4-9ff4-5477940e40b9">
<img width="595" alt="model-div-span-03" src="https://github.com/user-attachments/assets/488123e9-6e5a-4a37-b96e-ab870df0bd8c">

**Comment by jgm** [comment]
1. You can use `--no-highlight --lua-filter myhighlighter.lua` and write your own Lua filter `myhighlighter.lua` that highlights your code blocks using pygments or highlight.js or whatever tool you like.

2. You can grab the xml syntax definition for your language of interest from the [skylighting repository](https://github.com/jgm/skylighting/tree/master/skylighting-core/xml), modify it to your taste, and use it with pandoc with the `--syntax-definition` option.  Skylighting uses KDE highlighting definitions; you can look up the documentation for these which is linked from the README in the skylighting repository.  (E.g. for elm it ought to be possible to change the definition to use a different style for function arguments.)  Once you've made your improvement you might consider upstreaming it to KDE.

Both of these require some programming, I suppose, but nothing too complex.  I asked ChatGPT to do 1 and the result looks pretty credible (not tested):

```lua
-- A Lua filter for Pandoc that highlights code blocks using the external pygments program

-- Function to execute the pygments command and capture the output
local function highlight_with_pygments(code, lang)
  -- Define the command to run pygments
  local command = string.format('pygmentize -l %s -f html -O full,style=default', lang)
  -- Open a handle to the pygments process
  local handle = io.popen(command, 'w')
  -- Write the code to the process
  handle:write(code)
  handle:close()
  
  -- Capture the highlighted output
  local highlighted_code = io.popen(command):read('*a')
  
  return highlighted_code
end

-- The main function to process code blocks
function CodeBlock(elem)
  -- Check if the code block has a language specified
  if elem.classes[1] then
    -- Get the language of the code block
    local lang = elem.classes[1]
    
    -- Highlight the code using pygments
    local highlighted_code = highlight_with_pygments(elem.text, lang)
    
    -- Replace the code block content with the highlighted code
    return pandoc.RawBlock('html', highlighted_code)
  end
end
```

**Comment by citrusmunch** [comment]
The above AI answer didn't work immediately for me, but it was a decent starting point. I'm no lua expert, but I thought I'd share what I'm using now.

It takes about twice the time as the default pandoc highlighter; I'm converting 1365 markdown files in a little over 3 min on a nine-year-old machine. Perhaps my lua is inefficient, but this is good enough for me as a quick and dirty filter! I'm sure investing in a haskell version would be the way to go for anyone concerned with speed.

I'd love to hear any suggestions for improvement!

```lua
-- stx-highlight.lua

-- A Lua filter for Pandoc that highlights code blocks using the external
-- pygments program. Derived from:
-- https://github.com/jgm/pandoc/discussions/10115#discussioncomment-10433143

local string_format = string.format
local pandoc = pandoc
local RawBlock = pandoc.RawBlock

-- Pygmentize code for some lang
local function pygmentize(code, lang)

   -- write original code block contents to a temporary file
   -- https://stackoverflow.com/a/72273633
   local tmp_filename = os.tmpname()
   local tmp_handle = assert(io.open(tmp_filename, "w"))
   tmp_handle:write(code)
   tmp_handle:close()

   -- run the pygment program string as a process and read it's output
   local prog = string_format(
      "pygmentize -l %s -f html -O cssclass=stx-hl %s",
      lang,
      tmp_filename
   )
   local proc_handle = assert(io.popen(prog, "r"))
   local code_pygmentized = proc_handle:read("*a")

   -- clean up the process handle and temporary file
   proc_handle:close()
   os.remove(tmp_filename)

   return code_pygmentized
end

-- Replace code block content with pygmentized code if lang is specified.
function CodeBlock(elt)
   local lang = elt.classes[1]
   if lang then
      return RawBlock('html', pygmentize(elt.text, lang))
   end
end

```

**Reply by massifrg** [reply]
> I'd love to hear any suggestions for improvement!

Just a general tip to speed up lua code: define local variables for objects that are accessed many times.
So you may write:

```lua
local pandoc = pandoc
```

Global variables in Lua are keys of a global table `_G`; so `pandoc` is actually `_G.pandoc`.
With that line, you're creating a real local variable (on the stack?), preventing the resolution of the "pandoc" hash key in the `_G` table every time you refer to "pandoc".

Lua is interpreted, so reducing the number of hash resolutions has an effect on performance, when that number is high.

That has also a side effect on the `---@diagnostic disable-next-line: undefined-global`: you would need it just before `local pandoc = pandoc` at the top of your source, instead of having it in the middle of the code.
If you use VS Code, you can look [here](https://github.com/massifrg/pandoc-luals-annotations#prevent-undefined-global-warnings-with-luarcjson) to let the IDE know the global lua variables defined by pandoc.

In the same fashion, you may also write:

```lua
local string_format = string.format
local pandoc = pandoc
local RawBlock = pandoc.RawBlock
...
 -- run the pygment program string as a process and read it's output
   local prog = string_format(
      "pygmentize -l %s -f html -O cssclass=stx-hl %s",
      lang,
      tmp_filename
   )
...
-- Replace code block content with pygmentized code if lang is specified.
function CodeBlock(elt)
   local lang = assert(elt.classes[1])
   if lang then
      return RawBlock('html', pygmentize(elt.text, lang))
   end
end
```

It's a general tip: consider the tradeoff between code performance and readability case by case.

**Reply by citrusmunch** [reply]
Thanks for this! Unfortunately I think my main bottleneck is having to run pandoc separately for each markdown file, as there's no way to batch a directory afaik. Unless there's some background lua script caching going on then I think the speedup would be marginal in this case (which I'm also seeing experimentally).

Though, I'm going to edit my previous response to include your suggestions as a matter of style!

**Comment by badlydrawnrob** [comment]
@citrusmunch I don't understand your example file completely, but is that all the code you'd need to use a different highlighter? I've never written in Lua before.

My particular problem is a little more difficult, as I can only use something _that doesn't require_ a javascript file to work (like `highlight.js`). That's because it needs to work within [Anki spaced repetition flashcards](https://github.com/badlydrawnrob/anki) and the app doesn't play nicely with javascript.

I've already checked Pygments (I used to use that, in fact) and it doesn't offer much improvement over Haskell's Skylighting (at least not for Elm lang!) ... Be good to hear how your experiments have been going.

**Reply by citrusmunch** [reply]
Hi! Yeah, I'm only using make, pygmentize, and pandoc with some lua filters. My `template.html` generates pure html, though it's hardcoded with a link to my `style.css`. Which of course you would include in the anki(🧡) card styling. I'm using a modified version of [MVP.css](https://github.com/andybrewer/mvp) and a spliced in some pygments-specific stuff (and using that `cssclass` option from when I called pygmentize from lua). As I'm reviewing it all now, I'm realizing how much undocumented work I've done LOL 😅 

I wrote a Makefile that essentially uses pandoc like this (some flags are redundant):

```sh
pandoc --from=gfm --to=html --standalone \
    --css="css/style.css" \
    --template="template.html" \
    --lua-filter="pandoc-stuff/stx-highlight.lua" \
    --metadata document-css=false \
    "example/markdown/index.md" \
    --output="_site/index.html"
```

I'm quite happy with the output for my use case (short of wanting to tweak my css forever).

On the lua code: I'm basically shelling out to pygmentize and capturing that string to replace all the markdown code blocks with the new, pygmentized html. If elm isn't supported well enough, I would think you could use any program you want to generate satisfactory html and the lua filter here would work pretty identically.

I was tempted to generalize this stuff from my mini-project, but that's probably never going to happen, so I'll just publish that [here](https://codeberg.org/cnmne/bitburner-docs-gen) so you can have a working example haha

**tl;dr**: yes, I generate a bunch of standalone html with custom css and zero javascript!

**Comment by bpj** [comment]
@citrusmunch  You might want to usr [`pandoc.pipe`] rather than `os.popen`. An example is [here].

  [`pandoc.pipe`]: https://pandoc.org/lua-filters.html#pandoc.pipe
  [here]: https://pandoc.org/lua-filters.html#converting-abc-code-to-music-notation

Another example: a recipe for reading a YAML file (as data) in Lua. It presupposes that [yq] is in your path.

``` lua
-- Load a data file in one of several formats (default: yaml) via a yq pipe
local load_yq = function(file_name, format, ...)
  -- Default value for format
  if nil == format then
    format = "yaml"
  end
  -- Try to convert the data file to JSON and read it over a pandoc pipe.
  -- `-ojson` means output format JSON
  -- `-p` is input format, `format` is its argument
  -- Note the necessary empty string argument to `pipe` after the table
  --    with arguments to yq
  local ok, json = pcall(pandoc.pipe, 'yq', { '-ojson', '-p', format, ..., file_name }, "")
  -- If yq or the pipe failed `ok` is false and `json` is the error message
  if not (ok) then
    error("Failed to read " .. tostring(format) ..
      " '" .. tostring(file_name) .. "' over pipe with yq:\n" .. tostring(json))
  end
  -- Else (if `ok` is true) yq and the pipe succeeded and `json` contains
  --    the JSON string output by yq
  -- So try to decode it into data with pandoc.json
  local ok, data = pcall(pandoc.json.decode, json, false)
  -- If `ok` is true json.decode succeeded
  --    and `data` contains its return value
  if ok then
    return data
  end
  -- Else (if `ok` is false) json.decode failed,
  --    and `data` contains the error message
  return error("Failed to decode JSON: " .. tostring(data))
end
```
[yq]: https://github.com/mikefarah/yq
